### PR TITLE
csv-report - script, pipeline, samples

### PR DESCRIPTION
**Pull request by felickz**
Script and sample pipeline for querying the GHAzDO alert APIs and uploading a CSV report of an org's alerts for each repo as an artifact.

[Readme.md](https://github.com/microsoft/GHAzDO-Resources/blob/e7a702376423864752a27cc280af811b79a96730/src/csv-report/README.md)

See: [Sample Report](https://github.com/microsoft/GHAzDO-Resources/blob/ca3318bf7ccd02edd4a8ef72ec66081f0b162298/src/csv-report/ghazdo-report-20230905.1.csv)

### CODE REVIEW

File: src/csv-report/ghazdo-csv-report.ps1 [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+}
+$reportName = "ghazdo-report-$build.csv"
+
+if ($alertList.Count -gt 0) {
+    $alertList | Format-Table -AutoSize | Out-String | Write-Host
+    $alertList | Export-Csv -Path "$reportName" -NoTypeInformation -Force
+    Write-Host "##vso[artifact.upload artifactname=AdvancedSecurityReport]${env:BUILD_SOURCESDIRECTORY}/$reportName"
+}
+
+Write-Host "##vso[task.complete result=Succeeded;]DONE"
```
[the remark this answers is not in the document]
Comment (felickz) [reply src/csv-report/ghazdo-csv-report.ps1:150]: uses Write-Host. to output to Pipelines.

File: src/csv-report/ghazdo-csv-report.ps1 [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+$alertTypes = @("code", "secret", "dependency")
+$severityDays = @{
+    "critical" = 7
+    "high"     = 30
+    "medium"   = 90
+    "low"      = 180
+}
+
+# get list of projects in the Organization
+$url = "https://dev.azure.com/{0}/_apis/projects?api-version=6.0" -f $orgName
```
Comment (ncouraud) [inline src/csv-report/ghazdo-csv-report.ps1:46]: Is there a reason to not use the latest API versions?

File: src/csv-report/ghazdo-csv-report.ps1 [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+
+#loop through repo alert list
+[System.Collections.ArrayList]$alertList = @()
+foreach ($scan in $scans) {
+    $project = $scan.ProjectName
+    $repositoryName = $scan.RepoName
+    $repositoryId = $scan.RepoId
+    $alerts = $null
+    $parsedAlerts = $null
+    $url = "https://advsec.dev.azure.com/{0}/{1}/_apis/AdvancedSecurity/Repositories/{2}/alerts?useDatabaseProvider=true" -f $orgName, $project, $repositoryId
```
Comment (ncouraud) [inline src/csv-report/ghazdo-csv-report.ps1:89]: the useDatabaseProvider flag has been deprecated, and we shouldn't need it at this point.

File: src/csv-report/ghazdo-csv-report.ps1 [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+
+    # Send out warnings for any org/project/repo that we cannot access alerts for!
```
Comment (ncouraud) [inline src/csv-report/ghazdo-csv-report.ps1:91]: Should we rather just call the enablement API to determine if AdvSec is enabled?

**Review by ncouraud — COMMENTED** [review]
I think the one metacomment I might make is whether we should consider any changes to improve support for running both in and outside of pipelines. 

In a similar vein - would it make sense to support passing in a scope to grab alerts from via Param vs. the inline variable?

File: src/csv-report/ghazdo-csv-report.ps1 (hunk 4 above)
Comment (felickz) [reply src/csv-report/ghazdo-csv-report.ps1:91]: Added ✅ ! 

I like the idea that you could use the debug logs of this script to gather info about which repos are enabled or not

**Comment by felickz** [comment]
> I think the one metacomment I might make is whether we should consider any changes to improve support for running both in and outside of pipelines.
> 
> In a similar vein - would it make sense to support passing in a scope to grab alerts from via Param vs. the inline variable?

Script started from another template that was meant for AzDO pipeline, but i have modified it so you can run a bit easier with params.  Added a `scope` param to control org/proj/repo reporting

**Review by ncouraud — APPROVED** [review]

**Pull request merged**